### ISSUE TITLE
Extending (semi-)concrete structures

### ISSUE THREAD

**Issue by dselsam** [leanprover/lean3#921]
This issue comes courtesy of Nathaniel Thomas:

``` Lean
import algebra.field data.real
open algebra real

structure has_scalar [class] (F V : Type) :=
(smul : F → V → V)

structure vector_space [class] (F V : Type) extends has_scalar F V, field F :=
(idcomp   : ∀ x, smul one x = x)

structure binary_to_real [class] (V : Type) :=
(ip : V → V → ℝ)

structure real_inner_product [class] (V : Type) extends binary_to_real V, vector_space ℝ V :=
(positivity : ∀ (v : V), ((ip v v) ≥ 0))

set_option pp.all true
set_option pp.numerals false

definition positivity {V : Type} [inner : real_inner_product V] (v : V) 
  : real_inner_product.ip v v ≥ 0 :=
@real_inner_product.positivity V inner v
```

yields

```
error: type mismatch at definition 'positivity', has type
  ∀ (V : Type.{?M_1}) (inner : real_inner_product.{?M_1} V) (v : V),
    @ge.{1} real 11.source.to.has_le_3 (@real_inner_product.ip.{?M_1} V inner v v)
      (@zero.{1} real 11.source.to.has_zero_3)
but is expected to have type
  ∀ {V : Type.{?M_1}} [inner : real_inner_product.{?M_1} V] (v : V),
    @ge.{1} real 11.source.to.has_le_3 (@real_inner_product.ip.{?M_1} V inner v v)
      (@zero.{1} real (@real_inner_product.to.has_zero.{?M_1} V inner))
```

The issue is that by extending the semi-concrete structure `vector_space ℝ V`, we have introduced a new instance of `@zero ℝ` that has no actual definition, and in particular is not definitionally equal to the standard instance.

What is the recommended solution for this kind of problem?

**Comment by leodemoura** [comment]
I think the declaration of `vector_space` looks fishy. For example, if `F` is real, it is saying a vector space is a subclass of `real`. I think the definition should be:

``` lean
structure vector_space [class] (F V : Type) [s : field F] extends has_scalar F V :=
(idcomp   : ∀ x, smul one x = x)
```

`field F` is a parameter, i.e., a condition on `F`. The new definition says that to instantiate the class `vector_space` we need a `F` and it must be a field. The example works with this definition.

**Comment by dselsam** [comment]
Good point, thanks. I didn't read the set-up carefully. @nthomas103, do you agree?

So, fingers-crossed we never do need to extend a concrete structure that transitively extends something with a canonical implementation.

**Comment by leodemoura** [comment]
BTW, the class `real_inner_product` also looks fishy. It is saying a `real_inner_product` is a `vector_space`.

**Comment by dselsam** [comment]
I think that is just a confusing name, and that he meant it to refer to a "real inner-product space".

**Comment by leodemoura** [comment]
> So, fingers-crossed we never do need to extend a concrete structure that transitively extends something with a canonical implementation.

I don't see any problem. If it happens, we should not use type class resolution since we would not have a canonical instance anymore. Type class resolution works well for structures that have a canonical element (for the given parameters). In the definition above after we define `real_inner_product` we have two different ways to build a field/ring/has_zero/... of real.

**Comment by avigad** [comment]
Here are some thoughts. When we do abstract algebra, it will be convenient to use bundled structures:

   https://github.com/leanprover/lean/blob/master/library/algebra/bundled.lean

That way, we can write

``` lean
variable F : Field
variable V : VectorSpace F
```

instead of

``` lean
variable F : Type
variable [fieldF : field F]
include fieldF
variable V : Type
variable [vector_space_FV] : vector_space F V
include V
```

Similarly we can write

``` lean
variables F₁ F₂ : Field 
variables (V₁ : VectorSpace F₁) (V₂ : VectorSpace F₂)
```

rather than the much more verbose alternative. It is clunky to manage the "includes," i.e. turn them off or on manually depending on whether a theorem or definition needs them. Also, there is no need to make class inference search for the relevant instance when it can be bundled into the types. Similarly, I think we want a linear operator to be a bundled structure (function + properties) between (bundled) vector spaces,

``` lean
variable T : LinearOperator V₁ V₂
```

and so on,

Bundled structures are used extensively in Floris' category theory library, and we are similarly starting a group theory library based on them, so we can write

``` lean
variable G : Group
variables H K : Subgroup G
```

The thing is, when we do calculations with concrete vector spaces, I think we want the partially bundled versions. E.g. we may have

``` lean
x y : vector real n
c : real
```

and we want to write x + y or  smul c x  and have the system find the structure.

Then the solution should be to do what we currently do with groups: use the partially bundled versions for calculations in the structure, and then bundle them up when we want to define operations on groups, etc. 

The thing is, there are two things we can bundle or unbundle, namely, vector space and the field. At first, I thought we would want to keep the bundled structure for the underlying field, and only unbundle the vector space:

``` lean
variable F : Field
variable V : Type
variable [vector_space_F_V] : vector_space F V
```

But that doesn't seem to work: in the case above, given x : vector real n, there is no way for class inference to find the F (in this case, the bundled real field structure).

In short, it seems to me that for concrete structures, we want to use the approach you have described here (with the field structure and vector space structure discovered by class inference). When we move to algebra, if we want to bundle things up, we bundle up both at once.

I am just thinking out loud here, but if you think I am missing something, let me know.

**Comment by fpvandoorn** [comment]
That sounds right to me.

**Comment by leodemoura** [comment]
> rather than the much more verbose alternative. It is clunky to manage the "includes," i.e. turn them off or on manually depending on whether a theorem or definition needs them.

I think there is an easy solution for this. We can extend lean to automatically include any `[ ... ]` argument whenever all parameters of the type class have been used. That is, we would be able to write

``` lean
variables {F : Type} [f : field F]
variables  {V : Type}  [vs: vector_space F V]
```

Then, if we write 

``` lean
example (a b : F) : ...
```

Lean would automatically include `[f : field F]`. 

One possible drawback is a declaration that references `V` but not `F`. 

``` lean
example (v : V) : ...
```

I'm not sure how often this would happen in practice. we could still use the explicit `include` to address it.

We can also make the notation nicer by allowing "anonymous" `[...]` arguments. Then, we would be able to write

``` lean
variables {F : Type} [field F]
variables  {V : Type}  [vector_space F V]
```

This is almost as compact as the bundled version.

**Comment by fpvandoorn** [comment]
I like both suggestions. The anonymous `[...]` would be nice in general. We can still refer to the variable explicitly using ``field F``, right?

**Comment by avigad** [comment]
It sounds nice to me, but I think that in many situations (doing abstract algebra) is still better to use bundled structures. (Floris can probably confirm this.) 

In that case, the change would be nice, but it is not high priority, since for concrete structures we have been managing o.k. with the manual "include".

**Comment by leodemoura** [comment]
I think it would be nice to have a uniform library: less duplication, don't need to evaluate case-by-case which approach to use, ...

**Comment by avigad** [comment]
I don't think there is any reasonable way to avoid bundled structures. There is a big difference between reasoning in a particular algebraic structure (which is mostly what we have done so far in the standard library) and doing abstract algebra, where you have lots of structures floating around and operations on them. 

The difference is essentially that between writing f : Group -> Group and writing two functions, one on the types and one on the structures. But the problems get compounded quickly. Andrew and I gave up on using unbundled homomorphisms with subgroups; class inference started looping right away. Floris is the one who is doing "real" algebra, and he seems fairly convinced that we have to bundle things.

The good news is that I don't think the nonuniformity will cause any problems. We can easily apply unbundled theorems to bundled structures, and we can always bundle up an unbundled structure if we need to. It is generally clear whether we are working "in" a structure or dealing with operations on structures, so I think it will be clear which formulation we should use in any given situation.

**Comment by leodemoura** [comment]
Haitao managed to do group theory with the unbundled approach. He defined subgroups, proved non trivial lemmas, etc.
There is no mechanism for using a bundled theorem when working with unbundled ones.
The last thing I need is another issue asking for this kind of feature: bundle to unbundled.

**Comment by fpvandoorn** [comment]
This is an important issue. We should discuss this. I think it's infeasible to do everything unbundled, but I also have problems with the bundled approach.

I think we should move the discussion to a better place. I'll write my opinions more elaborately on lean-user.
